doc/generated_testcases_v2.yaml

```diff
@@ -0,0 +1,221 @@
+# and usually isn't required.
+visualizer: /visualizers/asy.py {name}
+
+# Optionally, a salt for generating the {seed} variables. Will be prepended to
+# the command being run.
+random_salt: abcd
+
+# We support three types of generators:
+# - Standalone files, like generators/a.cpp, generators/b.py, ..., which will
+#   be compiled and run the same way as submissions.
+# - Directories, like 
+#   - generators/gen/tree.cpp
+#   - generators/gen/lib.h
+#   This will be compiled and run the same was as directory validators. Build
+#   and run scripts may be used.
+# - 'implicit' generators whose dependencies are specified in the generators:
+#   key below. The dependencies may refer to any files relative to generators/.
+#   The generator will be build and run as if they formed a separate directory.
+#   E.g. the first example would be equivalent to
+#   - generators/tree/lib.py
+#   - generators/tree/tree.py
+#
+# When a generator is used as a command to generate a .in file, we first check
+# if the name is a key in the generators: dictionary below. If that is the case,
+# the corresponding generator is used. If not, we will use the file/directory
+# generator in the generator/ directory directly.
+generators:
+# A generator that depends on two files, lib.py and tree.py, directly in the
+# generators directory.
+  tree:
+    - lib.py
+    - tree.py
+# Another generator that also depends on the same lib.py.
+  graph:
+    - lib.py
+    - graph.py
+# This also works for other languages.
+  a:
+    - a.h
+    - a.cpp
+# Single-file generators may be specified, but can also be referred to as
+# b.cpp directly.
+  b:
+    - b.cpp
+# It is allowed, but not required, to explicitly list single-file generators
+# as well. It is allowed to reuse the same name, but introducing a new name is
+# also fine.
+  c.py:
+    - c.py
+    - lib.py
+
+data:
+  # Introduce the `sample` directory.
+  sample:
+    type: directory
+    solution: # empty to disable generating .ans files here
+    data:
+      1: stdout.py 1  # prints `1` to stdout, which is piped to 1.in
+
+# To indicate a manual testcase, specify it with an empty value.
+      2: 
+# Any key that matches the glob `*.in` is a manual testcase that will be copied
+# from the given directory into the target testcase. The given directory
+# must not start with a / and will be relative to generators/.
+      3: manual_cases/sample/3.in
+# Every testcase must be listed.
+# TOOLING: may allow this and print a warning.
+     #4:
+
+  secret:
+    type: directory
+    data:
+
+# Types of generator programs.
+      01: stdout.py  3   # stdout of the command is written to 01.in.
+      02: stdout.cpp 4   # c++ is compiled as for validators
+      03: dir 5          # directories are OK as for validators
+      04: tree 5         # keys from the global generators: dictionary may also be used.
+      05: a 5            # idem
+
+# The entire command is passed through python shlex.shlex.
```

Comment:
Would be nice to clarify this in non-python-specific terms: does this mean that `a b` is passed as a single token argument to the command `stdout.py`? What kind of quoting and escaping is possible/necessary.
Does any of this conflict with YAML escaping/quoting?

Reply:
The intent is simple: to allow shell-like quoting of arguments, so that `a b` will be passed as a single argument indeed.

So the processing will be:

1. YAML comments are ignored
2. YAML parses the string. it seems to me that escaping is rather minimal in YAML: ([stackoverflow source](https://stackoverflow.com/a/33000792/2716069))
    - put the entire value in single quotes. now use a double single quote `''` to get `'`, but nothing else
    - put the entire value in double quotes. Now you can use the normal `\` escapes: `\\`, `\n`, ...
    - Use fold mode: `>` followed by the (multi-line) string value on the next line(s).
  So with YAML you'd never quote only part of the string, and quoting the entire string in single quotes or fold mode doesn't interfere with what's inside.
3. Shell-parse the entire value. [shlex.split](https://docs.python.org/3.7/library/shlex.html?highlight=shlex%20split#shlex.split) calls [shlex.shlex](https://docs.python.org/3.7/library/shlex.html?highlight=shlex%20split#shlex.shlex) in POSIX mode and 'tries to be as close as possible to the POSIX shell parsing rules', with the actual rules [described here](https://docs.python.org/3.7/library/shlex.html#parsing-rules).

    I'm not sure whether or not we should set `punctuation_chars` to `True` or `False`. See [here](https://docs.python.org/3.7/library/shlex.html#improved-compatibility-with-shells) for the difference.


Anyway, you can use `'`, `"` and `\` within `"` as you're used from normal shells.

Reply:
Ok, but from a YAML what is the value of the string:
`stdout.py "a b"`
or
`stdout.py a b`
?

Reply:
YAML only uses quotes at the very start and end, so in this case the value will be `stdout.py "a b"`.

Right, we could execute the entire command through a shell, but that opens another can of worms. Now suddenly everything is _actual_ shell and you can use `|`, `;`, `&&`, `>`, `<`, ... . I definitely don't want that.
Also, it's better to be precise in the escaping we allow so we can have compatibility with windows.
I'm not sure how things work on Windows, but [subprocess.Popen](https://docs.python.org/3.7/library/subprocess.html?highlight=subprocess%20run#subprocess.Popen) has some info (see `shell=True`).